File: mobile/ios/Actnet/Sources/Models/InviteToken.swift
import Foundation

struct InviteToken: Identifiable, Hashable {
    var id: String { token }
    let token: String
    let serverUrl: URL
    let serverName: String
    let inviterDid: String?
    let postOnboardingRedirect: String?
    /// Operator's privacy policy URL, resolved by the core as part of invite
    /// validation (same source as `GET /v1/info`) — no separate server call.
    /// nil when none is configured or the value is blank. Onboarding screens
    /// show the link only when non-nil.
    let privacyPolicyURL: URL?

    /// Parse a go.theavalanche.net invite URL and validate the token with the server.
    static func from(url: URL) async throws -> InviteToken {
        // Extract token from path: /i/<token> (short form; "invite" still
        // accepted for any older links).
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard let action = pathComponents.first, action == "i" || action == "invite",
              pathComponents.count >= 2 else {
            throw InviteError.invalidURL
        }
        let token = pathComponents[1]
        return try await from(token: token)
    }

    /// Validate a raw base64url token string with the server.
    static func from(token: String) async throws -> InviteToken {
        let info = try validateInvite(token: token)
        guard let serverUrl = URL(string: info.serverUrl) else {
            throw InviteError.invalidServerUrl
        }
        return InviteToken(
            token: token,
            serverUrl: serverUrl,
            serverName: info.serverName,
            inviterDid: info.inviterDid,
            postOnboardingRedirect: info.postOnboardingRedirect,
            privacyPolicyURL: info.privacyPolicyUrl
                .flatMap { $0.isEmpty ? nil : URL(string: $0) }
        )
    }
}

enum InviteError: LocalizedError {
    case invalidURL
    case invalidServerUrl

    var errorDescription: String? {
        switch self {
        case .invalidURL: "Invalid invite link"
        case .invalidServerUrl: "Invalid server URL in invite"
        }
    }
}

File: mobile/android/app/src/main/kotlin/Models/InviteToken.kt
package net.theavalanche.app

import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.app_core.validateInvite

/// Parsed and server-validated invite token.
/// Mirrors InviteToken.swift — iOS is the reference implementation.
data class InviteToken(
    val token: String,
    val serverUrl: String,
    val serverName: String,
    val inviterDid: String?,
    val postOnboardingRedirect: String?,
) {
    // Mirrors Swift's `var id: String { token }` from Identifiable conformance.
    val id: String get() = token

    companion object {
        /// Parse a go.theavalanche.net invite URL and validate the token with the server.
        /// Mirrors InviteToken.from(url:) — path must be /i/<token> or /invite/<token>.
        /// (Named fromUrl because Kotlin, unlike Swift, has no argument labels to
        /// distinguish this from the raw-token overload.)
        suspend fun fromUrl(url: String): InviteToken {
            val uri = Uri.parse(url)
            val segments = uri.pathSegments.filter { it.isNotEmpty() }
            val action = segments.firstOrNull()
            if (action == null || (action != "i" && action != "invite") || segments.size < 2) {
                throw InviteError.InvalidURL
            }
            val token = segments[1]
            return fromToken(token)
        }

        /// Validate a raw base64url token string with the server.
        /// Calls the UniFFI top-level validateInvite() function on a background thread.
        suspend fun fromToken(token: String): InviteToken {
            val info = withContext(Dispatchers.IO) {
                validateInvite(token = token)
            }
            return InviteToken(
                token = token,
                serverUrl = info.serverUrl,
                serverName = info.serverName,
                inviterDid = info.inviterDid,
                postOnboardingRedirect = info.postOnboardingRedirect,
            )
        }
    }
}

sealed class InviteError(message: String) : Exception(message) {
    /// The URL did not match the expected /i/<token> or /invite/<token> pattern.
    object InvalidURL : InviteError("Invalid invite link")

    /// The server URL embedded in the invite info could not be parsed.
    object InvalidServerUrl : InviteError("Invalid server URL in invite")
}
